runtime/devnet/src/extensions/mod.rs

```diff
@@ -0,0 +1,1160 @@
+					message.encode().into(),
+				)
+				.refund_surplus()
+				.deposit_asset(assets.into(), beneficiary)
+				.build();
+			(dest, message)
+		},
+	};
+	// TODO: revisit to replace with signed contract origin
+	let origin: RuntimeOrigin = RawOrigin::Root.into();
+	// Generate runtime call to dispatch.
+	let call = RuntimeCall::PolkadotXcm(pallet_xcm::Call::send {
+		dest: Box::new(dest),
+		message: Box::new(VersionedXcm::V4(message)),
+	});
+	dispatch_call::<T, E>(env, call, origin, LOG_PREFIX)
+}
+
+// Converts a `DispatchError` to a `u32` status code based on the version of the API the contract uses.
+// The contract calling the chain extension can convert the status code to the descriptive `Error`.
+//
+// For `Error` see `pop_primitives::<version>::error::Error`.
+//
+// The error encoding can vary per version, allowing for flexible and backward-compatible error handling.
+// As a result, contracts maintain compatibility across different versions of the runtime.
+//
+// # Parameters
+//
+// - `error`: The `DispatchError` encountered during contract execution.
+// - `version`: The version of the chain extension, used to determine the known errors.
+pub(crate) fn convert_to_status_code(error: DispatchError, version: u8) -> u32 {
+	// "UnknownFunctionId" and "DecodingFailed" are mapped to specific errors in the API and will
+	// never change.
+	let mut encoded_error = match error {
+		DispatchError::Other("UnknownFunctionId") => Vec::from([254u8, 0, 0, 0]),
+		DispatchError::Other("DecodingFailed") => Vec::from([255u8, 0, 0, 0]),
+		_ => error.encode(),
+	};
+	// Resize the encoded value to 4 bytes in order to decode the value in a u32 (4 bytes).
+	encoded_error.resize(4, 0);
+	let mut encoded_error = encoded_error.try_into().expect("qid, resized to 4 bytes line above");
+	match version {
+		// If an unknown variant of the `DispatchError` is detected the error needs to be converted
+		// into the encoded value of `Error::Other`. This conversion is performed by shifting the bytes one
+		// position forward (discarding the last byte as it is not used) and setting the first byte to the
+		// encoded value of `Other` (0u8). This ensures the error is correctly categorized as an `Other`
+		// variant which provides all the necessary information to debug which error occurred in the runtime.
+		//
+		// Byte layout explanation:
+		// - Byte 0: index of the variant within `Error`
+		// - Byte 1:
+		//   - Must be zero for `UNIT_ERRORS`.
+		//   - Represents the nested error in `SINGLE_NESTED_ERRORS`.
+		//   - Represents the first level of nesting in `DOUBLE_NESTED_ERRORS`.
+		// - Byte 2:
+		//   - Represents the second level of nesting in `DOUBLE_NESTED_ERRORS`.
+		// - Byte 3:
+		//   - Unused or represents further nested information.
+		0 => v0::error::handle_unknown_error(&mut encoded_error),
+		_ => encoded_error = [254, 0, 0, 0],
+	}
+	u32::from_le_bytes(encoded_error)
+}
+
+#[derive(Debug)]
+pub enum FuncId {
```

Comment:
Missing doc comments for enum definition and variants

Reply:
Thanks, see comment "docs: add docs to `pub` code"